### PR TITLE
Arwen WASM VM integration.

### PR DESCRIPTION
**Pull request by sasurobert**
Integration and tests of Arwen WASM VM

### CODE REVIEW

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+func TestVmDeployWithCPUCalculateAndExecute(t *testing.T) {
+	runWASMVMBenchmark(t, "./cpucalculate_arwen.wasm", 100, 8000)
+}
+
+func TestVmDeployWithStringConcatAndExecute(t *testing.T) {
+	runWASMVMBenchmark(t, "./stringconcat_arwen.wasm", 100, 10000)
+}
+
+func runWASMVMBenchmark(t *testing.T, fileSC string, numRun int, testingValue uint64) {
```
Comment (iulianpascalau) [inline integrationTests/vm/arwen/arwenVM_test.go]: this benchmark shall be done through golang's built in benchmark system

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	fmt.Printf("time elapsed to process topup %s \n", elapsedTime.String())
+	assert.Nil(t, err)
+
+	_, err = accnts.Commit()
+	assert.Nil(t, err)
+
+	aliceNonce++
+
+	start = time.Now()
+	for i := 0; i < 10000; i++ {
```
Comment (iulianpascalau) [inline integrationTests/vm/arwen/arwenVM_test.go]: This is no longer a test. Should split into a test and a benchmark

File: process/smartContract/scDataGetter.go [hunk 3]
```diff
@@ -15,24 +16,33 @@ var maxGasValue = big.NewInt(math.MaxInt64)
...
 	}
 
 	return &scDataGetter{
-		vm: vm,
+		vmContainer: vmContainer,
 	}, nil
 }
 
+func (scdg *scDataGetter) getVMFromAddress(scAddress []byte) (vmcommon.VMExecutionHandler, error) {
+	vmType := scAddress[hooks.NumInitCharactersForScAddress-hooks.VMTypeLen : hooks.NumInitCharactersForScAddress]
```
Comment (iulianpascalau) [inline process/smartContract/scDataGetter.go]: Why not creating an exported func in hooks package that will output the vmType as byte slice? Something like
```GO
vmType := hooks.ComputeVmType(scAddress)
```
and hooks.ComputeVmType will have the code as found on line 38.

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	vm.TestAccount(
+		t,
+		accnts,
+		senderAddressBytes,
+		senderNonce+1,
+		expectedBalance)
+}
+
+func Benchmark_VmDeployWithFibbonacciAndExecute(b *testing.B) {
```
Comment (iulianpascalau) [inline integrationTests/vm/arwen/arwenVM_test.go]: We should decide how to change these tests-with-manual-evaluation to something else (not benchmarks nor tests).

File: process/smartContract/scDataGetter.go (hunk 3 above)
Comment (sasurobert) [reply process/smartContract/scDataGetter.go]: done

File: integrationTests/vm/arwen/arwenVM_test.go (hunk 4 above)
Comment (sasurobert) [reply integrationTests/vm/arwen/arwenVM_test.go]: done

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			GasLimit:  5000,
+			Data:      "_main",
+			Signature: nil,
+			Challenge: nil,
+		}
+
+		startTime := time.Now()
+		err = txProc.ProcessTransaction(tx, round)
+		elapsedTime := time.Since(startTime)
+		fmt.Printf("time elapsed full process %s \n", elapsedTime.String())
```
Comment (AdoAdoAdo) [inline integrationTests/vm/arwen/arwenVM_test.go]: I would not do print inside a benchmark. 
There is no need to calculate the time for processTransaction manually, you can benchmark just ProcessTransaction separately if you are not interested in commit as well

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		assert.Nil(tb, err)
+
+		_, err = accnts.Commit()
+		assert.Nil(tb, err)
+
+		aliceNonce++
+	}
+}
+
+func TestVmDeployWithTransferAndExecuteERC20(t *testing.T) {
```
Comment (AdoAdoAdo) [inline integrationTests/vm/arwen/arwenVM_test.go]: this test is too long, you could extract some functions, maybe for creating types of transactions used in here.

File: integrationTests/vm/testInitializer.go [hunk 7]
```diff
@@ -289,3 +284,11 @@ func GetAccountsBalance(addrBytes []byte, accnts state.AccountsAdapter) *big.Int
 
 	return shardAccnt.Balance
 }
+
+func GetIntValueFromSC(accnts state.AccountsAdapter, scAddressBytes []byte, funcName string, args ...[]byte) *big.Int {
+	vmContainer, _ := CreateVMsContainerAndBlockchainHook(accnts)
+	scgd, _ := smartContract.NewSCDataGetter(vmContainer)
+
+	returnedVals, _ := scgd.Get(scAddressBytes, funcName, args...)
+	return big.NewInt(0).SetBytes(returnedVals)
```
Comment (AdoAdoAdo) [inline integrationTests/vm/testInitializer.go]: leave an empty line before return

File: process/smartContract/scDataGetter.go [hunk 8]
```diff
@@ -15,24 +16,33 @@ var maxGasValue = big.NewInt(math.MaxInt64)
...
 	}, nil
 }
 
+func (scdg *scDataGetter) getVMFromAddress(scAddress []byte) (vmcommon.VMExecutionHandler, error) {
+	vmType := hooks.VMTypeFromAddressBytes(scAddress)
+	vm, err := scdg.vmContainer.Get(vmType)
+	if err != nil {
+		return nil, err
+	}
+	return vm, nil
```
Comment (AdoAdoAdo) [inline process/smartContract/scDataGetter.go]: leave an empty line before return

File: integrationTests/vm/testInitializer.go (hunk 7 above)
Comment (sasurobert) [reply integrationTests/vm/testInitializer.go]: done

File: process/smartContract/scDataGetter.go (hunk 8 above)
Comment (sasurobert) [reply process/smartContract/scDataGetter.go]: done

File: integrationTests/vm/arwen/arwenVM_test.go (hunk 5 above)
Comment (sasurobert) [reply integrationTests/vm/arwen/arwenVM_test.go]: deleted and changed. I still like the print variant better.

File: integrationTests/vm/arwen/arwenVM_test.go (hunk 6 above)
Comment (sasurobert) [reply integrationTests/vm/arwen/arwenVM_test.go]: refactored some part.

**Pull request merged**